eth/protocols/eth/receipt.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	"errors"
+	"io"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type receiptForNetwork types.Receipt
+
+func (r *receiptForNetwork) DecodeRLP(s *rlp.Stream) error {
+	kind, size, err := s.Kind()
+	if err != nil {
+		return err
+	}
+	if kind != rlp.List {
+		return errors.New("invalid receipt")
+	}
+	if size == 4 {
```

Comment:
In the RLP format, the size is always given in bytes, even for lists. There is no way to know the number of items in a list before decoding the whole list. So this check doesn't work, and it's why I suggested to just encode *all* receipt types, even legacy ones, in the format `[type, status-or-root, gasUsed, logs]`.

My suggestion here would be something like:

```go
_, err := s.List()
if err != nil {
    return err
}
defer s.ListEnd()
rtype, err := s.Uint8()
if err != nil {
    return err
}
status, err := s.Bytes()
if err != nil {
    return err
}
gasUsed, err := s.Uint64()
if err != nil {
    return nil
}
var logs []*types.Log
if err := s.Decode(&logs); err != ni {
    return err
}

receipt := types.Receipt{
    Type: rtype,
    GasUsed: gasUsed,
    Logs: logs,
}
if len(status) == 32 {
    receipt.Root = status
} else if len(status) == 8 {
    receipt.Status = binary.BigEndian.Uint64(status)
}
```
```